sql-plugin/src/main/scala/com/nvidia/spark/rapids/JoinGatherer.scala

```diff
@@ -670,4 +675,133 @@ case class MultiJoinGather(left: JoinGatherer, right: JoinGatherer) extends Join
   }
 
   override def toString: String = s"MULTI-GATHER $left and $right"
-}
\ No newline at end of file
+}
+
+/**
+ * JoinGatherer for existence join
+ * Existence join generates an `exists` boolean column with `true` or `false` in it,
+ * then appends it to the `output` columns. It does not shrink or expand left table.
+ * Just provides the `exists` column for the next following operator(usually filter).
+ *
+ * e.g.:
+ * <code>
+ * select * from left_table where
+ *   left_table.column_0 >= 3
+ *   or
+ *   exists (select * from right_table where left_table.column_1 < right_table.column_1)
+ *
+ * Explanation of this sql is:
+ *
+ * Filter(left_table.column_0 >= 3 or `exists`)
+ *   Existence_join // generate `exists` column, do not shrink or expand the rows of left table
+ *     left_table
+ *     right_table
+ * </code>
+ */
+class JoinGathererForExistenceJoin(
+    private val gatherMap: LazySpillableGatherMap,
+    private val data: LazySpillableColumnarBatch,
+    boundsCheckPolicy: OutOfBoundsPolicy)
+    extends JoinGathererImpl(gatherMap, data, boundsCheckPolicy) {
+
+  // How much of the gather map we have output so far
+  private var gatheredUpTo: Long = 0
+
+  // Note: equals to the rows of the left table
+  private val totalRows: Long = data.numRows
+
+  override def isDone: Boolean =
+    gatheredUpTo >= totalRows
+
+  override def numRowsLeft: Long = totalRows - gatheredUpTo
+
+  override def toString: String = {
+    s"GATHERER FOR EXISTENCE JOIN $gatheredUpTo/$totalRows $gatherMap $data"
+  }
+
+  /**
+   * Existence join does not shrink or expand,
+   * it only adds an `exists` boolean column to left table according to the gather map
+   */
+  override def gatherNext(n: Int): ColumnarBatch = {
+    val start = gatheredUpTo
+    assert((start + n) <= totalRows)
+    val types = GpuColumnVector.extractTypes(data.getBatch)
+    val batch = data.getBatch
+
+    val subTableCbTmp = withResource(GpuColumnVector.from(batch)) { table =>
+      if(start == 0 && n == totalRows) {
+        // no need to split
+        GpuColumnVector.from(table, types)
+      } else {
+        withResource(table.contiguousSplit(start.toInt, (start + n).toInt)) { tableSplits =>
+          assert(tableSplits.length >= 2)
+          GpuColumnVector.from(tableSplits(1).getTable, types)
+        }
+      }
+    }
+
+    val ret = withResource(subTableCbTmp) { subTableCb =>
+      val subExistsCvTmp = withResource(gatherMap.toColumnView(0, gatherMap.getRowCount.toInt)) {
+        gatherView =>
+          // `exists.numRows` == `batch.numRows`,
+          //  with true or false in it indicating if the row is gathered
+          withResource(genExistsColumn(gatherView, batch.numRows())) { exists =>
```

Comment:
Personally I perfer to generate the column of existence according to the `subTableCbTmp`, not the whole batch. Then there is no need to split it next. Instead you can append the generated 'existence' column to the sub table directly.

Reply:
Just as @firestarman mentioned, this may lead to a lot of duplicated runs of full-scale scatter. IIUC, if we split the gather map as left table, we need to transform the global offset to the local offset  before scattering. Perhaps there is another approach, just cache the exist column for reuse over mini batches.

Reply:
Emit the batch directly as the `exists` boolean column is relatively small.
It's the same as `GpuHashJoin` does, no need to split now.